src/cpp/session/modules/SessionPackages.R

```diff
@@ -1226,79 +772,692 @@ if (identical(as.character(Sys.info()["sysname"]), "Darwin") &&
+      source <- sub("Bioconductor", "BioC", source, fixed = TRUE)
+      
+      # Check for R-universe remotes, and format them specially.
+      pattern <- "^https?://(.*)\\.r-universe\\.dev/?$"
+      m <- regexec(pattern, source)
+      matches <- regmatches(source, m)[[1L]]
+      if (length(matches))
+         source <- sprintf("R-universe [%s]", matches[[2L]])
+      
+      # Check for PPM remorts, and format them specially.
+      if (identical(pkgDesc[["Repository"]], "RSPM"))
+      {
+         repos <- pkgDesc[["RemoteRepos"]]
+         if (!is.null(repos))
+         {
+            ppm <- if (grepl("https://p3m.dev/", repos, fixed = TRUE))
+               "p3m.dev"
+            else if (grepl("https://packagemanager.posit.co/", repos, fixed = TRUE))
+               "P3M"
+            else
+               "PPM"
+            
+            snapshot <- basename(repos)
+            name <- basename(dirname(repos))
+            source <- sprintf("%s [%s/%s]", ppm, name, snapshot)
+         }
+      }
+      
+      return(source)
    }
    
-   # Write various files out
-   
-   # NOTE: write.dcf mangles whitespace so we manually construct
-   # the text we wish to write out
-   DESCRIPTION <- lapply(DESCRIPTION, function(field) {
-      paste(field, collapse = "\n    ")
-   })
-   
-   names <- names(DESCRIPTION)
-   values <- unlist(DESCRIPTION)
-   text <- paste(names, ": ", values, sep = "", collapse = "\n")
-   cat(text, file = file.path(packageDirectory, "DESCRIPTION"), sep = "\n")
+   # Handle other remotes. Start by mapping the RemoteType field
+   # into a "pretty" version of the type.
+   aliases <- list(
+      bioc         = "Bioconductor",
+      bioconductor = "Bioconductor",
+      bitbucket    = "Bitbucket",
+      cran         = "CRAN",
+      git2r        = "Git",
+      github       = "GitHub",
+      gitlab       = "GitLab",
+      local        = "Local",
+      repository   = "Repository",
+      standard     = "Repository",
+      url          = "URL",
+      xgit         = "Git"
+   )
    
-   cat(NAMESPACE, file = file.path(packageDirectory, "NAMESPACE"), sep = "\n")
+   type <- .rs.nullCoalesce(aliases[[rtype]], rtype)
    
-   RprojPath <- file.path(
-      packageDirectory,
-      paste(packageName, ".Rproj", sep = "")
-   )
+   # Check for packages installed from e.g. hosted Git repositories.
+   user <- pkgDesc[["RemoteUsername"]]
+   repo <- pkgDesc[["RemoteRepo"]]
+   if (!is.null(user) && !is.null(repo))
+   {
+      source <- sprintf("%s [%s/%s]", type, user, repo)
+      return(source)
+   }
    
-   if (!.Call("rs_writeProjectFile", RprojPath, PACKAGE = "(embedding)"))
-      return(.rs.error("Failed to create package .Rproj file"))
+   # If we couldn't infer any more details, just use the type.
+   type
    
-   # Ensure new packages get AutoAppendNewLine + StripTrailingWhitespace
-   Rproj <- readLines(RprojPath)
+})
+
+.rs.addFunction("inferPackageBrowseUrl", function(pkgDesc)
```

Comment:
New code; this links to a landing page of the repository from where this package was retrieved. Normally this will be a CRAN / PPM landing page, but could be a GitHub URL for packages installed from GitHub, and similar for other providers.